Applications/Spire/Source/Ui/InfoPanel.cpp

```diff
@@ -0,0 +1,151 @@
+  }
+
+  const auto& ERROR_IMAGE() {
+    static auto image =
+      imageFromSvg(":/Icons/info_panel/error.svg", SEVERITY_ICON_SIZE());
+    return image;
+  }
+}
+
+InfoPanel::InfoPanel(QString message, QWidget& parent)
+  : InfoPanel(Severity::INFO, std::move(message), parent) {}
+
+InfoPanel::InfoPanel(Severity severity, QString message, QWidget& parent)
+    : QWidget(&parent),
+      m_severity(severity),
+      m_message(std::move(message)) {
+  setAttribute(Qt::WA_DeleteOnClose);
+  auto image = [&] {
+    if(m_severity == Severity::INFO) {
+      return INFO_IMAGE();
+    } else if(m_severity == Severity::SUCCESS) {
+      return SUCCESS_IMAGE();
+    } else if(m_severity == Severity::WARNING) {
+      return WARNING_IMAGE();
+    } else if(m_severity == Severity::ERROR) {
+      return ERROR_IMAGE();
+    }
+    return QImage();
+  }();
+  auto severity_icon = new Icon(image);
+  severity_icon->setFixedSize(SEVERITY_ICON_SIZE());
+  update_style(*severity_icon, [&] (auto& styles) {
+    if(m_severity == Severity::INFO) {
+      styles.get(Any() || Hover()).set(Fill(QColor(0x4392D6)));
+    } else if(m_severity == Severity::SUCCESS) {
+      styles.get(Any() || Hover()).set(Fill(QColor(0x26BF4A)));
+    } else if(m_severity == Severity::WARNING) {
+      styles.get(Any() || Hover()).set(Fill(QColor(0xFFBB00)));
+    } else if(m_severity == Severity::ERROR) {
+      styles.get(Any() || Hover()).set(Fill(QColor(0xE63F44)));
+    }
+  });
+  auto message_area_layout = make_hbox_layout();
+  message_area_layout->addWidget(severity_icon, 0, Qt::AlignTop);
+  message_area_layout->addSpacing(scale_width(4));
+  m_label = make_text_area_label(m_message);
+  message_area_layout->addWidget(m_label);
+  auto left_area_layout = make_vbox_layout();
+  left_area_layout->addSpacing(scale_height(1));
+  left_area_layout->addLayout(message_area_layout);
+  auto close_button = make_icon_button(
+    imageFromSvg(":/Icons/info_panel/close.svg", scale(16, 16)));
+  close_button->setFixedSize(scale(16, 16));
+  close_button->connect_click_signal([=] { close(); });
+  auto layout = make_hbox_layout(this);
+  layout->addLayout(left_area_layout);
+  layout->addSpacing(scale_width(8));
+  layout->addWidget(close_button, 0, Qt::AlignTop);
+  m_panel = new OverlayPanel(*this, parent);
+  update_style(*m_panel, [] (auto& styles) {
+    styles.get(Any()).
+      set(horizontal_padding(scale_width(8))).
+      set(vertical_padding(scale_height(8)));
+  });
+  m_panel->set_closed_on_focus_out(false);
+  m_panel->set_is_draggable(false);
+  m_panel->installEventFilter(this);
+  on_message_style();
+  connect_style_signal(*m_label,
+    std::bind_front(&InfoPanel::on_message_style, this));
+}
+
+InfoPanel::Severity InfoPanel::get_severity() const {
+  return m_severity;
+}
+
+const QString& InfoPanel::get_message() const {
+  return m_message;
+}
+
+bool InfoPanel::eventFilter(QObject* watched, QEvent* event) {
+  if(event->type() == QEvent::Show) {
+    show();
```

Comment:
Can you explain why this ```show()``` is needed? Doesn't this widget get shown automatically by ```m_panel```?

Reply:
Yes, `m_panel` can automatically bring the widget up. This is redundant. I will remove it.